### PR TITLE
feat: expose prometheus metrics

### PR DESCRIPTION
**Pull request by samsja**
see internal docs here : https://docs.google.com/document/d/17NqWFh8Q9bkOkHw3et-_ghWkl_mISu2AjE5LC3dsp3M/edit

- [x] add prometheus endpoint on each pods
- [x] add prometheus endpoint on gateway for all gateway type
- [x]  add prometheus endpoint for the head
- [x] expose a subset of core metrics for pods
- [x] expose a subset of core metrics for the gateway
- [x]  expose a subset of core metrics for the head
- [x] Allow `Executor` to expose their own metrics
- [x] refactor `Executor` init to be able to pass prometheus registry
- [x] make sure that the monitoring server is closed properly when quitting: blocked, need to find a smart way
- [x] print monitoring adress with as the other adress
- [x] add support for docker compose 

further work after this pr:

Implement the other metrics that we want to expose https://github.com/jina-ai/jina/issues/4589
Design a grafana dashboard https://github.com/jina-ai/jina/issues/4588
documentation of all the monitoring feature https://github.com/jina-ai/jina/issues/4587
k8s support https://github.com/jina-ai/jina/issues/4590

### CODE REVIEW

File: jina/serve/runtimes/worker/__init__.py [hunk 1]
```diff
@@ -3,13 +3,14 @@
 import multiprocessing
 import threading
 from abc import ABC
-from typing import Optional, Union, List
+from typing import List, Optional, Union
 
 import grpc
+from prometheus_client import start_http_server
```
[the remark this answers is not in the document]
Comment (samsja) [reply jina/serve/runtimes/worker/__init__.py:9]: oh yes sure

File: jina/serve/runtimes/worker/__init__.py [hunk 2]
```diff
@@ -36,6 +37,16 @@ def __init__(
         self._data_request_handler = DataRequestHandler(args, self.logger)
 
     async def async_setup(self):
+        """
+        Wait for the GRPC and Monitoring (if needed) servers to start
+        """
+        await self._async_setup_grpc_server()
```
[the remark this answers is not in the document]
Comment (samsja) [reply jina/serve/runtimes/worker/__init__.py:43]: I will check

[the remark this answers is not in the document]
Comment (samsja) [reply jina/serve/runtimes/worker/__init__.py:42]: the pr is still in early progress rn

File: tests/integration/monitoring/test_pod.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class DummyExecutor(Executor):
+    @requests
+    def foo(self, docs, **kwargs):
+        print('here')
+
+
+def test_enable_monitoring_deployment():
+    port1 = 8089
```
Comment (JoanFM) [inline tests/integration/monitoring/test_pod.py:17]: these ports need to be randomized

File: tests/integration/monitoring/test_pod.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        uses=DummyExecutor, monitoring=True, port_monitoring=port2
+    ):
+        for port in [port1, port2]:
+            resp = req.get(f'http://localhost:{port}/')
+            assert resp.status_code == 200
+
+
+@pytest.mark.parametrize('protocol', ['websocket', 'grpc', 'http'])
+def test_enable_monitoring_gateway(protocol):
+    port0 = 8088
```
Comment (JoanFM) [inline tests/integration/monitoring/test_pod.py:30]: same here

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/integration/monitoring/test_pod.py (hunk 3 above)
Comment (samsja) [reply tests/integration/monitoring/test_pod.py:17]: using `helper.random_port()` ?

**Comment by hanxiao** [comment]
may minorly related https://github.com/jina-ai/jina/issues/4561#issue-1185415797

File: jina/serve/runtimes/gateway/__init__.py [hunk 5]
```diff
@@ -3,13 +3,17 @@
...
 
 
-class GatewayRuntime(AsyncNewLoopRuntime, ABC):
+class GatewayRuntime(AsyncNewLoopRuntime, MonitoringMixin, ABC):
     """
     The Runtime from which the GatewayRuntimes need to inherit
     """
 
+    def _setup_monitoring(self):
+        MonitoringMixin._setup_monitoring_server(self)
```
Comment (JoanFM) [inline jina/serve/runtimes/gateway/__init__.py:15]: u should use `self._setup_monitoring_server()` or why is this not added directly in the Mixin?

File: jina/serve/runtimes/request_handlers/data_request_handler.py [hunk 6]
```diff
@@ -10,28 +10,40 @@
...
+        :param metrics_registry: optional metrics registry for prometheus used if we need to expose metrics from the executor of from the data request handler
         :param kwargs: extra keyword arguments
         """
         super().__init__()
         self.args = args
         self.args.parallel = self.args.shards
         self.logger = logger
         self._is_closed = False
-        self._load_executor()
+        self._load_executor(metrics_registry)
```
Comment (JoanFM) [inline jina/serve/runtimes/request_handlers/data_request_handler.py:40]: Since when `Executor` requires or has a `metrics_registry` attribute? It is wrong for an external entity to add attributes to an object.

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/serve/runtimes/gateway/__init__.py (hunk 5 above)
Comment (samsja) [reply jina/serve/runtimes/gateway/__init__.py:15]: The idea is that we could want to define some metrics on the gateway level and we can't do it the `MonitoringMixin` because it is shared for gateway and worker.

**Comment by samsja** [comment]
block by https://github.com/jina-ai/jina/pull/4581 for now

File: jina/orchestrate/flow/base.py [hunk 7]
```diff
@@ -1512,6 +1520,26 @@ def host(self, value: str):
...
+
+    @property
+    def port_monitoring(self) -> int:
+        """Return if the monitoring is enabled
+        .. # noqa: DAR201
+        """
+        if GATEWAY_NAME in self._deployment_nodes:
+            return self[GATEWAY_NAME].args.port_monitoring
+        else:
+            return __default_port_monitoring__
```
Comment (JoanFM) [inline jina/orchestrate/flow/base.py:1541]: I do not understand this logic. Who is guaranteeing that there is something in this default port?

File: jina/parsers/orchestrate/runtimes/head.py [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import argparse
 
+from jina import helper
```
Comment (JoanFM) [inline jina/parsers/orchestrate/runtimes/head.py:3]: this does not seem needed

File: jina/resources/k8s/template/deployment.yml [hunk 9]
```diff
@@ -32,6 +32,7 @@ spec:
           args: {args}
           ports:
             - containerPort: {port}
+            - containerPort: {port-monitoring}
```
Comment (JoanFM) [inline jina/resources/k8s/template/deployment.yml:35]: if this is not yet handled, I would suggest to remove it for now

File: jina/serve/networking.py [hunk 10]
```diff
@@ -298,6 +305,17 @@ def __init__(
             compression.lower(), grpc.Compression.NoCompression
         )
 
+        if metrics_registry:
+            from prometheus_client import Summary
+
+            self._summary_time = Summary(
+                'sending_request_seconds',
+                'Time spent between sending a request to the pod and receiving the response',
```
Comment (JoanFM) [inline jina/serve/networking.py:313]:
```suggestion
                'Time spent between sending a request to the Pod and receiving the response',
```

File: jina/serve/networking.py [hunk 11]
```diff
@@ -477,26 +495,29 @@ async def task_wrapper(requests, stubs, endpoint):
                     request_type = type(requests[0])
                     if request_type == DataRequest and len(requests) == 1:
 
-                        call_result = stubs[0].process_single_data(
+                        call_result = stubs[0].process_single_data(  # todo investigate
```
Comment (JoanFM) [inline jina/serve/networking.py:498]: investigate what?

File: jina/serve/runtimes/gateway/__init__.py [hunk 12]
```diff
@@ -3,6 +3,7 @@
 from jina.serve.networking import GrpcConnectionPool
 from jina.serve.runtimes.asyncio import AsyncNewLoopRuntime
 from jina.serve.runtimes.gateway.graph.topology_graph import TopologyGraph
+from jina.serve.runtimes.mixin import MonitoringMixin
```
Comment (JoanFM) [inline jina/serve/runtimes/gateway/__init__.py:6]: this is not used

File: jina/serve/runtimes/gateway/graph/topology_graph.py [hunk 13]
```diff
@@ -79,7 +79,10 @@ async def _wait_previous_and_send(
                             DataRequestHandler.reduce_requests(self.parts_to_send)
                         ]
 
-                    resp, metadata = await connection_pool.send_requests_once(
+                    (
+                        resp,
+                        metadata,
+                    ) = await connection_pool.send_requests_once(  # todo investigate
```
Comment (JoanFM) [inline jina/serve/runtimes/gateway/graph/topology_graph.py:85]: investigate what?

File: jina/serve/runtimes/mixin.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
+import threading
+from typing import TYPE_CHECKING
```
Comment (JoanFM) [inline jina/serve/runtimes/mixin.py:2]: this file should not be called `mixin` but `monitoring`

File: jina/serve/runtimes/mixin.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    )
+    t = threading.Thread(target=httpd.serve_forever)
+    t.daemon = True
+    t.start()
+
+    return httpd
+
+
+def _close_monitoring_server(httpd):
+    t = threading.Thread(target=httpd.shutdown())
```
Comment (JoanFM) [inline jina/serve/runtimes/mixin.py:32]: I do not think there is need to do it in a Thread?

File: jina/serve/runtimes/mixin.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    :param registry: prometheus registry that will feed to http server with metrics
+    :param addr: address on which to expose
+    :return: httpd server
+    """
+    from wsgiref.simple_server import make_server
+
+    from prometheus_client import make_wsgi_app
+    from prometheus_client.exposition import ThreadingWSGIServer, _SilentHandler
+
+    app = make_wsgi_app(registry)
```
Comment (JoanFM) [inline jina/serve/runtimes/mixin.py:20]: Why are we doing this? This was handled by `prometheus` before. What changed?

**Review by JoanFM — CHANGES_REQUESTED** [review]
Besides all comments, please check that no uneeded imports are added

**Comment by JoanFM** [comment]
Another task that needs to be added is to benchmark the perfromance with and without monitoring enabled

File: jina/serve/runtimes/worker/__init__.py [hunk 17]
```diff
@@ -32,7 +33,29 @@ def __init__(
         """
         super().__init__(args, cancel_event, **kwargs)
 
+        # Keep this initialization order, otherwise readiness check is not valid
+        self._data_request_handler = DataRequestHandler(
```
Comment (JoanFM) [inline jina/serve/runtimes/worker/__init__.py:37]: Review this part, as this has changed in a recent commit to master, I guess a bad merge happened here

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/orchestrate/flow/base.py (hunk 7 above)
Comment (samsja) [reply jina/orchestrate/flow/base.py:1541]: Then what should I return if the gateway_name is not in the deployment node ?

> Who is guaranteeing that there is something in this default port?
Nothing guarantee that, it is just saying on which port the monitoring would be serve if the monitoring is activaed. There is another attribute to say if it is activated or not

File: jina/resources/k8s/template/deployment.yml (hunk 9 above)
Comment (samsja) [reply jina/resources/k8s/template/deployment.yml:35]: my bad I thought that I had removed that part

**Comment by samsja** [comment]
> Another task that needs to be added is to benchmark the performance with and without monitoring enabled

done https://github.com/jina-ai/jina/issues/4591

File: jina/parsers/orchestrate/pod.py [hunk 18]
```diff
@@ -97,3 +97,18 @@ def mixin_pod_parser(parser):
...
+        default=False,
+        help='If set, spawn an http server with a prometheus endpoint to expose metrics',
+    )
+
+    gp.add_argument(
+        '--port-monitoring',
+        type=int,
+        default=__default_port_monitoring__,  # default prometheus server port
+        dest='port_monitoring',
+        help='The port on which the prometheus server is exposed, default a random port between [49152, 65535]',
```
Comment (jacobowitz) [inline jina/parsers/orchestrate/pod.py:113]: it defaults to __default_port_monitoring__ no? Although it is right that locally it will be random if you run multiple things

File: jina/serve/runtimes/gateway/graph/topology_graph.py [hunk 19]
```diff
@@ -79,7 +79,7 @@ async def _wait_previous_and_send(
                             DataRequestHandler.reduce_requests(self.parts_to_send)
                         ]
 
-                    resp, metadata = await connection_pool.send_requests_once(
+                    (resp, metadata,) = await connection_pool.send_requests_once(
```
Comment (jacobowitz) [inline jina/serve/runtimes/gateway/graph/topology_graph.py:82]: why this change?

**Review by jacobowitz — CHANGES_REQUESTED** [review]
having metrics will be a great feature! Only two minor comments from my side

File: jina/serve/runtimes/gateway/graph/topology_graph.py (hunk 19 above)
Comment (samsja) [reply jina/serve/runtimes/gateway/graph/topology_graph.py:82]: No idea, I would say it happened which black or something. Will fix it

**Review by jacobowitz — APPROVED** [review]
great! :+1:

**Review by JoanFM — APPROVED** [review]
